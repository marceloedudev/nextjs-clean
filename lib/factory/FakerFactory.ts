import FakerAddress from '../faker/FakerAddress'
import faker from 'faker-br'
import FakerData from '../faker/FakerData'
import FakerCommerce from '../faker/FakerCommerce'
import FakerInternet from '../faker/FakerInternet'
import FakerName from '../faker/FakerName'
import FakerPhone from '../faker/FakerPhone'
import FakerRandom from '../faker/FakerRandom'
import FakerLorem from '../faker/FakerLorem'

class FakerFactory {
  createAddress() {
    return new FakerAddress(faker)
  }

  createCommerce() {
    return new FakerCommerce(faker)
  }

  createData() {
    return new FakerData(faker)
  }

  createInternet() {
    return new FakerInternet(faker)
  }

  createLorem() {
    return new FakerLorem(faker)
  }

  createName() {
    return new FakerName(faker)
  }

  createPhone() {
    return new FakerPhone(faker)
  }

  createRandom() {
    return new FakerRandom(faker)
  }
}

export default FakerFactory
