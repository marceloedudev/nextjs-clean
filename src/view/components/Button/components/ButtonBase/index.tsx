import React, { useCallback } from 'react'

import { Container } from './styles'

export type IButtonBase = {
  onClick?: (event) => void
  disabled?: boolean
  type?: 'button' | 'reset' | undefined
  color?: string
  label?: string
  textColor?: string
}

export const ButtonBase: React.FC<IButtonBase> = ({
  onClick,
  disabled,
  type,
  color,
  label,
  textColor
}) => {
  const onMouseClickHandler = useCallback(
    (event) => {
      if (onClick) {
        onClick(event)
      }
    },
    [onClick]
  )

  return (
    <Container
      onClick={onMouseClickHandler}
      disabled={disabled}
      data-testid="button"
      type={type || 'button'}
      color={color ?? '#6c5ce7'}
      textColor={textColor ?? '#FFF'}
    >
      <span className="content">{label}</span>
    </Container>
  )
}
