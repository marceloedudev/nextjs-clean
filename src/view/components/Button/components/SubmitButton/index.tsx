import React, { useCallback } from 'react'

import { Container } from './styles'
import { LoadingIcon } from '@/resources/icons'

export type ISubmitButtonBase = {
  onClick?: (event) => void
  disabled?: boolean
  loading?: boolean
  color?: string
  label?: string
  textColor?: string
}

export const SubmitButtonBase: React.FC<ISubmitButtonBase> = ({
  onClick,
  disabled,
  loading,
  color,
  label,
  textColor
}) => {
  const onMouseClickHandler = useCallback(
    (event) => {
      if (onClick) {
        onClick(event)
      }
    },
    [onClick]
  )

  return (
    <Container
      onClick={onMouseClickHandler}
      disabled={disabled || loading}
      data-testid="submit"
      type={'submit'}
      color={color}
      textColor={textColor || '#FFF'}
    >
      <span className="content">
        {loading ? (
          <div className="icon">
            <LoadingIcon />
          </div>
        ) : null}
        {label}
      </span>
    </Container>
  )
}
