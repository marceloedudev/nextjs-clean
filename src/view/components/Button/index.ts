export * from './components/SubmitButton'
export * from './components/ButtonBase'
