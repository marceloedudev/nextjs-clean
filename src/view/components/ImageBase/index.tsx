import React from 'react'
import Image from 'next/image'

export type ImageBaseProps = {
  width: number
  height: number
  src: string
  alt?: string
}

const ImageBase = ({ width, height, src, alt }: ImageBaseProps) => {
  return <Image width={width} height={height} src={src} alt={alt} />
}

export default ImageBase
