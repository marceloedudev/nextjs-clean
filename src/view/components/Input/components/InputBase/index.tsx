import { Container, HelperContainer, Input, Label } from './styles'

import React from 'react'

export interface IFloatingLabelInputProps {
  icon: JSX.Element
}

export interface IFloatingLabelInput
  extends React.InputHTMLAttributes<HTMLInputElement> {
  name: string
  label: string
  helperText?: string
  state?: any
  setState?: any
  inputProps?: IFloatingLabelInputProps
  autoComplete?: string | 'undefined'
  ref?: any
}

export const InputBase: React.FC<IFloatingLabelInput> = React.forwardRef(
  (props: IFloatingLabelInput, ref) => {
    const {
      name,
      label,
      helperText = '',
      inputProps,
      state,
      setState,
      autoComplete,
      onChange,
      onFocus,
      ...restProps
    } = props

    const defaultRef = React.useRef()
    const resolvedRef: any = ref || defaultRef

    const [values, setValues] = React.useState({
      active: false,
      focusActive: false
    })

    const onMouseBlurHandler = React.useCallback(
      (event) => {
        setValues((oldState) => ({
          ...oldState,
          active: event.target.value.length !== 0,
          focusActive: false
        }))
      },
      [setValues]
    )

    const onMouseFocusHandler = React.useCallback(
      (event) => {
        if (onFocus) {
          onFocus(event)
        }
        setValues((oldState) => ({
          ...oldState,
          active: true,
          focusActive: true
        }))
      },
      [onFocus, setValues]
    )

    const onMouseChangeHandler = React.useCallback(
      (event) => {
        if (onChange) {
          onChange(event)
        }
        setState({ ...state, [event.target.name]: event.target.value })
      },
      [onChange, state, setState]
    )

    return (
      <div>
        <Container
          active={values.active}
          helperActive={helperText && helperText?.length}
          focusActive={values.focusActive}
        >
          <Label
            htmlFor={name}
            active={values.active}
            helperActive={helperText && helperText?.length}
            focusActive={values.focusActive}
            data-testid={`inputbase-label`}
          >
            {label}
          </Label>

          <Input
            name={name}
            active={values.active}
            helperActive={helperText && helperText?.length}
            focusActive={values.focusActive}
            onBlur={onMouseBlurHandler}
            onFocus={onMouseFocusHandler}
            onChange={onMouseChangeHandler}
            ref={resolvedRef}
            autoComplete={autoComplete || 'off'}
            data-testid={name}
            {...restProps}
          />
          {inputProps && inputProps.icon && (
            <div
              data-testid={`inputbase-icon`}
              className="inputbase-icon-container"
            >
              <div className="inputbase-icon-content">{inputProps.icon}</div>
            </div>
          )}
        </Container>

        <HelperContainer>
          {helperText && <p data-testid={`${name}-helper`}>{helperText}</p>}
        </HelperContainer>
      </div>
    )
  }
)
