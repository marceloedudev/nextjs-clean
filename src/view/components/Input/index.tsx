export * from './components/InputBase'
export * from './components/PasswordInputBase'
