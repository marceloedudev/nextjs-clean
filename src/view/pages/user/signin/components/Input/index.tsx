import { InputBase } from '@/view/components/Input'
import React from 'react'
import { useSignIn } from '../../contexts/useSignIn'

export interface IInput {
  name: string
  type: string
  label: string
  icon?: JSX.Element
  helperText?: string
  autoComplete?: string
}

const Input: React.FC<IInput> = ({
  name,
  type,
  label,
  icon,
  helperText,
  autoComplete
}) => {
  const { state, actions } = useSignIn()
  return (
    <InputBase
      name={name}
      type={type}
      label={label}
      inputProps={{
        icon: { ...icon }
      }}
      helperText={helperText}
      autoComplete={autoComplete}
      state={state.data}
      setState={actions.setData}
    />
  )
}

export default Input
