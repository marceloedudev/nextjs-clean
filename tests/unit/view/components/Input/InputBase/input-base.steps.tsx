import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { InputBase } from '@/view/components/Input'
import { expect, beforeAll, jest } from '@jest/globals'
import { defineFeature, loadFeature } from 'jest-cucumber'
import path from 'path'
import FakerFactory from '../../../../../../lib/factory/FakerFactory'

const pathMakeFeature = path.resolve(path.join(__dirname, 'input-base.feature'))

const feature = loadFeature(pathMakeFeature)

const makeComponent = (params) => {
  const { fieldName, fieldLabel, onFocus, onChange } = params
  const Demo = () => {
    const [data, setData] = React.useState({
      testeName: ''
    })

    return (
      <>
        <div>{data.testeName}</div>
        <InputBase
          name={fieldName}
          label={fieldLabel}
          state={data}
          setState={setData}
          onFocus={onFocus}
          onChange={onChange}
        />
      </>
    )
  }

  return render(<Demo />)
}

defineFeature(feature, (test) => {
  let fakerLorem

  beforeAll(() => {
    const fakerFactory = new FakerFactory()
    fakerLorem = fakerFactory.createLorem()
  })

  test('updating value property correctly', ({ when, then }) => {
    let instanceComponent

    const onChange = jest.fn()
    const onBlur = jest.fn()
    const onFocus = jest.fn()

    when('mount input component', () => {
      const fieldLabel = 'Input Test'
      const fieldName = 'pass-demo2'
      instanceComponent = makeComponent({
        fieldName,
        fieldLabel,
        onFocus,
        onChange
      })
    })

    then('compare the value property', () => {
      const { container } = instanceComponent
      const inputEl = container.querySelector('input')
      fireEvent.input(inputEl, {
        inputType: 'deleteContentBackward'
      })
      expect(inputEl.value).toBe('')
    })

    then("check if it didn't call onBlur event", () => {
      expect(onBlur).not.toHaveBeenCalled()
    })
  })

  test('snapshot component', ({ when, then }) => {
    let instanceComponent

    when('to mount the component', () => {
      instanceComponent = makeComponent({
        fieldName: 'testeName',
        fieldLabel: `label input test`
      })
    })

    then('render perfectly', () => {
      expect(instanceComponent.container).toMatchSnapshot()
    })
  })

  test('updating name properties correctly', ({ when, then }) => {
    let instanceComponent

    when('to mount the component', () => {
      instanceComponent = makeComponent({
        fieldName: 'testName',
        fieldLabel: `label input test`
      })
    })

    then('compare the name property', () => {
      const inputEl = instanceComponent.container.querySelector('input')
      expect(inputEl.getAttribute('name')).toBe('testName')
    })
  })

  test('verify change state and setState', ({ given, when, then }) => {
    let instanceComponent
    const onChange = jest.fn()

    given('component assembled with state change on screen', () => {
      const onChangeHandler = () => {
        onChange()
      }
      const fieldLabel = fakerLorem.paragraph()
      instanceComponent = makeComponent({
        fieldName: 'testeName',
        fieldLabel: `${fieldLabel}`,
        onChange: onChangeHandler
      })
    })

    when('typing something in the field', () => {
      const inputEl = instanceComponent.container.querySelector('input')
      fireEvent.change(inputEl, { target: { value: 'test text' } })
    })

    then('check the number of times called', () => {
      expect(onChange).toHaveBeenCalledTimes(1)
    })

    then('the text on the screen should change', () => {
      expect(screen.getByText(/test text/i)).toBeTruthy()
    })
  })
})
