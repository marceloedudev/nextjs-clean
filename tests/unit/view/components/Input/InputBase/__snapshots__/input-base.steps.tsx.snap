// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Base snapshot component 1`] = `
<div>
  <div />
  <div>
    <div
      class="css-1vv8m27"
    >
      <label
        class="css-30bnu8"
        data-testid="inputbase-label"
        for="testeName"
      >
        label input test
      </label>
      <input
        autocomplete="off"
        class="css-1vbdnyf"
        data-testid="testeName"
        name="testeName"
      />
    </div>
    <div
      class="css-kli5ab"
    />
  </div>
</div>
`;
